File: src/fragment/ItemTypeFragment.ts
import { Text } from "pixi.js";
import { StructEndecBuilder, PrimitiveEndecs } from 'KEndec';
import Fragment, { FragmentType, register } from "./Fragment";
import { Identifier } from "~/util";


export default class ItemTypeFragment extends Fragment {
    item: string;

    constructor(item: string) {
        super();

        this.item = item;
    }

    override toString(): string {
        return this.item;
    }

    override type(): FragmentType<ItemTypeFragment> {
        return ITEM_TYPE;
    }
}

const ITEM_TYPE = register("Item_type", 0x2266aa,     
    StructEndecBuilder.of1(
        Identifier.ENDEC
            .xmap((ident) => ident.toString(), Identifier.of)
            .fieldOf("entityType", (fragment: ItemTypeFragment) => fragment.item),
        (item) => new ItemTypeFragment(item)
    )
);

File: site/src/jsMain/kotlin/me/maplesyrum/tricksterStudio/spell/fragment/ItemTypeFragment.kt
package me.maplesyrum.tricksterStudio.spell.fragment

import me.maplesyrum.tricksterStudio.Identifier
import tree.maple.kendec.StructEndec
import tree.maple.kendec.impl.StructEndecBuilder


class ItemTypeFragment(val item: Identifier) : Fragment() {
    override fun type(): FragmentType<*> {
        return FragmentType.ITEM_TYPE
    }

    override fun toString(): String {
        return item.toString()
    }

    companion object {
        val ENDEC: StructEndec<ItemTypeFragment> = StructEndecBuilder.of(
            Identifier.ENDEC
            .fieldOf("entityType") { fragment: ItemTypeFragment -> fragment.item }
        ) { item -> ItemTypeFragment(item) }
    }
}
